File: attractor/pipeline/edge_selector.py
"""5-step edge selection algorithm."""

from __future__ import annotations

from .conditions import evaluate
from .context import PipelineContext
from .graph import Edge


def select_edge(
    edges: list[Edge],
    outcome: str = "",
    preferred_label: str = "",
    suggested_next_ids: list[str] | None = None,
    context: PipelineContext | None = None,
) -> Edge | None:
    """Select the next edge using the 5-step priority algorithm.

    Steps:
    1. Filter edges with conditions, keeping those that evaluate to True
    2. Match preferred_label against edge labels
    3. Match suggested_next_ids against edge targets
    4. Pick highest weight
    5. Lexical tiebreak by target node ID
    """
    if not edges:
        return None

    ctx = context or PipelineContext()
    suggested = suggested_next_ids or []

    # Step 1: Evaluate conditions — keep edges where condition is True
    eligible: list[Edge] = []
    for edge in edges:
        if edge.condition:
            if evaluate(edge.condition, ctx, outcome, preferred_label):
                eligible.append(edge)
        else:
            eligible.append(edge)

    if not eligible:
        return None

    # Step 2: Match preferred_label
    if preferred_label:
        label_matches = [e for e in eligible if e.label == preferred_label]
        if label_matches:
            eligible = label_matches

    # Step 3: Match suggested_next_ids
    if suggested and len(eligible) > 1:
        suggested_matches = [e for e in eligible if e.target in suggested]
        if suggested_matches:
            eligible = suggested_matches

    # Step 4: Highest weight
    if len(eligible) > 1:
        max_weight = max(e.weight for e in eligible)
        eligible = [e for e in eligible if e.weight == max_weight]

    # Step 5: Lexical tiebreak
    if len(eligible) > 1:
        eligible.sort(key=lambda e: e.target)

    return eligible[0]

File: internal/pipeline/edgeselector.go
package pipeline

import "sort"

// SelectEdge picks the next edge using the 5-step priority algorithm:
//  1. keep edges whose condition evaluates true (unconditional edges always kept)
//  2. prefer edges matching preferredLabel
//  3. prefer edges whose target is in suggestedNextIDs
//  4. keep the highest-weight edges
//  5. lexical tiebreak by target id
func SelectEdge(edges []*Edge, outcome, preferredLabel string, suggestedNextIDs []string, ctx *PipelineContext) *Edge {
	if len(edges) == 0 {
		return nil
	}
	if ctx == nil {
		ctx = NewPipelineContext(nil)
	}

	// Step 1: conditions
	var eligible []*Edge
	for _, e := range edges {
		if e.Condition != "" {
			if ok, err := EvaluateCondition(e.Condition, ctx, outcome, preferredLabel); err == nil && ok {
				eligible = append(eligible, e)
			}
		} else {
			eligible = append(eligible, e)
		}
	}
	if len(eligible) == 0 {
		return nil
	}

	// Step 2: preferred_label
	if preferredLabel != "" {
		var matches []*Edge
		for _, e := range eligible {
			if e.Label == preferredLabel {
				matches = append(matches, e)
			}
		}
		if len(matches) > 0 {
			eligible = matches
		}
	}

	// Step 3: suggested_next_ids
	if len(suggestedNextIDs) > 0 && len(eligible) > 1 {
		suggested := map[string]bool{}
		for _, s := range suggestedNextIDs {
			suggested[s] = true
		}
		var matches []*Edge
		for _, e := range eligible {
			if suggested[e.Target] {
				matches = append(matches, e)
			}
		}
		if len(matches) > 0 {
			eligible = matches
		}
	}

	// Step 4: highest weight
	if len(eligible) > 1 {
		maxW := eligible[0].Weight
		for _, e := range eligible {
			if e.Weight > maxW {
				maxW = e.Weight
			}
		}
		var matches []*Edge
		for _, e := range eligible {
			if e.Weight == maxW {
				matches = append(matches, e)
			}
		}
		eligible = matches
	}

	// Step 5: lexical tiebreak
	if len(eligible) > 1 {
		sort.SliceStable(eligible, func(i, j int) bool { return eligible[i].Target < eligible[j].Target })
	}

	return eligible[0]
}
